File: solutions/1648-minimum-insertions-to-balance-a-parentheses-string/minimum-insertions-to-balance-a-parentheses-string.py
class Solution:
    def minInsertions(self, s: str) -> int:
        open_counts = 0
        insertions = 0

        i = 0
        while i < len(s):
            if s[i] == '(':
                open_counts += 1
            elif i + 1 < len(s) and s[i + 1] == ')':
                if open_counts > 0:
                    open_counts -= 1
                else:
                    insertions += 1
                i += 1  # Skip the next ')'
            else:
                if open_counts > 0:
                    open_counts -= 1
                    insertions += 1
                else:
                    insertions += 2
            i += 1

        return insertions + 2 * open_counts

File: solutions/1648-minimum-insertions-to-balance-a-parentheses-string/minimum-insertions-to-balance-a-parentheses-string.cpp
#include <string>
using namespace std;

class Solution {
public:
    int minInsertions(string s) {
        int openCounts = 0; // Tracks unmatched '('
        int insertions = 0; // Tracks required insertions

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                openCounts++;
            } else if (i + 1 < s.size() && s[i + 1] == ')') {
                if (openCounts > 0) openCounts--;
                else insertions++;
                i++; // Skip the next ')'
            } else {
                if (openCounts > 0) {
                    openCounts--;
                    insertions++;
                } else {
                    insertions += 2;
                }
            }
        }

        return insertions + 2 * openCounts;
    }
};